File: Codes/Bingran Li/HashTable/136. Single Number.java
public class Solution {
    public int singleNumber(int[] nums) {
        
       // someone's solution: bitwise,linear time without extra memory
       // N^N=0;0^N=N

       int result=0;
       for(int i: nums){
           result = result^i;
       }
       return result;
       
       // solution 2: hashset
       /*
       HashSet <Integer>mySet = new HashSet<Integer>();
       for(int i: nums){
           if(!mySet.add(i)) 
                mySet.remove(i);
       }
        return mySet.iterator().next(); 
        */
    }
}

File: Codes/Kan Zheng/C++/Hashtable/136. Single Number.cpp
class Solution {
public:

    int singleNumber(vector<int>& nums) {
        /*Use extra memory
        unordered_set <int> set;
        for(int i=0;i<nums.size();i++)
        {
            if(set.find(nums[i])==set.end())
                set.insert(nums[i]);
            else
                set.erase(nums[i]);
        }
        return *(set.begin());
        */
        
        //Use XOR
        int result=0;
        for(int i=0;i<nums.size();i++)
        {
            result ^= nums[i];
        }
        return result;
        
    }
};